src/rustdoc-json-types/lib.rs

```diff
@@ -52,11 +52,67 @@ pub struct Crate {
     pub paths: HashMap<Id, ItemSummary>,
     /// Maps `crate_id` of items to a crate name and html_root_url if it exists.
     pub external_crates: HashMap<u32, ExternalCrate>,
+    /// Information about the target for which this documentation was generated
+    pub target: Target,
     /// A single version number to be used in the future when making backwards incompatible changes
     /// to the JSON output.
     pub format_version: u32,
 }
 
+/// Information about a target
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Target {
+    /// The target triple for which this documentation was generated
+    pub triple: String,
+    /// A list of `#[target_feature]` which exist for this target, along with their status in this
+    /// compiler session
+    pub target_features: Vec<TargetFeature>,
+}
+
+/// Information about a target feature.
+///
+/// Rust target features are used to influence code generation, especially around selecting
+/// instructions which are not universally supported by the target architecture.
+///
+/// Target features are commonly enabled by the [`#[target_feature]` attribute][1] to influence code
+/// generation for a particular function, and less commonly enabled by compiler options like
+/// `-Ctarget-feature` or `-Ctarget-cpu`. Targets themselves automatically enable certain target
+/// features by default, for example because the target's ABI specification requires saving specific
+/// registers which only exist in an architectural extension.
+///
+/// Target features can imply other target features: for example, x86-64 `avx2` implies `avx`, and
+/// aarch64 `sve2` implies `sve`, since both of these architectural extensions depend on their
+/// predecessors.
+///
+/// Target features can be probed at compile time by [`#[cfg(target_feature)]`][2] or `cfg!(…)`
+/// conditional compilation to determine whether a target feature is enabled in a particular
+/// context.
+///
+/// [1]: https://doc.rust-lang.org/stable/reference/attributes/codegen.html#the-target_feature-attribute
+/// [2]: https://doc.rust-lang.org/reference/conditional-compilation.html#target_feature
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TargetFeature {
+    /// The name of this target feature.
+    pub name: String,
+    /// Other target features which are implied by this target feature, if any.
+    pub implies_features: Vec<String>,
```

Comment:
I assume:
- order doesn't matter
- each name may appear at most once
- features are guaranteed to form a DAG (directed acyclic graph), here including a forest of such DAGs

If all this is true, perhaps we should note it in the doc comment. If not, we should definitely note that it isn't true :)

The first two items can be made obvious by making this set instead of a `Vec`, though I'm ambivalent on whether that's an improvement or not. Your call!

Reply:
This all _should_ be true, but it's not ensured by the underlying data structures. `rustc_target::target_features::ImpliedFeatures` is just a `&'static [&'static str]`. Implying nonexistent features or forming a cycle wouldn't cause problems unless that feature were enabled.

If we're providing these guarantees, I'd be inclined to have `librustdoc` verify and `panic!()` if `rustc_target` tries to violate them. Is adding this check (and failure mode) desirable? If not, should we document these expectations with weaker language?

Separately, I note that `rustdoc-json-types` does not currently use any sets, only `Vec`s. It uses and re-exports `rustc_hash::FxHashMap` for maps. Should I re-export `FxHashSet` too?

Reply:
I'll defer to rustdoc JSON maintainers on whether `librustdoc` should verify the guarantees explicitly.

From my point of view, it's fine to expose either a `Vec` or `FxHashSet` here.

If the properties are expected but not able to be guaranteed, I'd also be in favor of documenting them with weaker language.

Reply:
I don't have strong feelings either way on `Vec` vs set type for places where order doesn't matter. If the expectation (from docs or from context) is that order is irrelevant, my preference would be "whatever is faster to iterate over" so probably slight preference toward a `Vec` overall :)